src/H5Cmpio.c

```diff
@@ -1010,18 +1012,21 @@ H5C__collective_write(H5F_t *f)
                 HGOTO_ERROR(H5E_CACHE, H5E_NOTFOUND, FAIL, "can't retrieve skip list item")
 
             /* Set up array position */
+            HDassert(entry_ptr->type);
             addrs[i] = entry_ptr->addr;
             sizes[i] = entry_ptr->size;
             bufs[i]  = entry_ptr->image_ptr;
-            types[i] = H5FD_MEM_DEFAULT;
+            types[i] = entry_ptr->type->mem_type;
+            if (types[i] != types[i - 1])
+                identical_types = FALSE;
 
             /* Advance to next node & array location */
             node = H5SL_next(node);
             i++;
         } /* end while */
 
         /* Optimization for vector I/O */
-        if (count > 1)
+        if (count > 1 && identical_types)
```

Comment:
I'm not sure if this identical types optimization here really buys us anything. We already have the array, might as well pass it to the FD layer. I suppose you could delay allocating the types array until the main loop is finished, and do a second pass if the types are not identical.

Reply:
I thought the idea was to tell the vector I/O implementation that it doesn't need to do any processing on the rest of the types array after the first element? Not that there's likely to be any measurable overhead in either case, but it's essentially free to do here, whereas if we don't the vector I/O code will walk the types array.

Reply:
@fortnern I've removed that particular piece of code for now. I still think it'll perform slightly better for large metadata writes, but I'll keep an eye on it in the future.